### PR TITLE
[GLUTEN-2956][VL] Support Spark NullType

### PR DESCRIPTION
**Pull request by PHILO-HE**
## What changes were proposed in this pull request?

Spark's NullType can be mapped to substrait's nothing type and velox's UnknownType. And we also need some code changes in Velox to support this type.

(Fixes: \#2956)

## How was this patch tested?

Existing Spark UTs & new UTs to be added.

### CODE REVIEW

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxValidator.scala [hunk 1]
```diff
@@ -51,6 +51,7 @@ class Validator extends ValidatorApi {
       case _: DecimalType =>
       case _: DateType =>
       case _: TimestampType =>
+      case _: NullType =>
```
Comment (winningsix) [inline backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxValidator.scala:54]: nit: Refactor this a little bit

```
        case _: BinaryType | BooleanType | ByteType | DateType | DoubleType | FloatType | IntegerType | LongType | 
             ShortType | StringType | TimestampType | NullType =>
        case _: DecimalType =>
        case _: ArrayType  =>
        case _: MapType =>
        case _: StructType =>
        case _ =>
```

Comment (PHILO-HE) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxValidator.scala]: Refactored. Thanks!

**Review by FelixYBW — APPROVED** [review]
passed customer test

**Comment by PHILO-HE** [comment]
As it depends on a patch for velox, let's wait for velox upstream contribution and velox code rebase.

File: cpp/velox/shuffle/VeloxShuffleWriter.cc [hunk 2]
```diff
@@ -1220,6 +1223,17 @@ arrow::Status VeloxShuffleWriter::splitFixedWidthValueBuffer(const velox::RowVec
         case arrow::MapType::type_id:
         case arrow::ListType::type_id:
           break;
+        case arrow::NullType::type_id: {
+          std::shared_ptr<arrow::Buffer> validityBuffer = nullptr;
+          ARROW_RETURN_NOT_OK(pool_->allocate(validityBuffer, arrow::bit_util::BytesForBits(newSize)));
```
Comment (zhouyuan) [inline cpp/velox/shuffle/VeloxShuffleWriter.cc:1228]: this piece of code needs to be updated to work with main branch

**Comment by FelixYBW** [comment]
@PHILO-HE  what's the status of this PR? Is the PR submited to upstream Velox?

**Comment by PHILO-HE** [comment]
> @PHILO-HE what's the status of this PR? Is the PR submited to upstream Velox?

Hi @FelixYBW, we have submitted required PR to velox upstream, but still not merged. I will push the progress soon. Thanks!

**Comment by FelixYBW** [comment]
> equired PR to velox upstream, but still not

which PR in velox?

**Comment by PHILO-HE** [comment]
> > equired PR to velox upstream, but still not
> 
> which PR in velox?

Hi @FelixYBW, we have 3 velox PRs for this. One is merged, two are still in progress.
https://github.com/facebookincubator/velox/pull/8724
https://github.com/facebookincubator/velox/pull/6996
https://github.com/facebookincubator/velox/pull/8769

**Review by zhouyuan — APPROVED** [review]
👍

**Comment by FelixYBW** [comment]
@PHILO-HE 8769 还没merge

**Comment by PHILO-HE** [comment]
> @PHILO-HE 8769 isn't merged yet. is it impact anything?

Hi @FelixYBW, the remaining one velox pr doesn't impact on the basic functionality of NullType and our CI can pass with oap/velox main branch. So we just merged this PR.

**Pull request merged**